src/appengine/handlers/external_update.py

```diff
@@ -0,0 +1,124 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""External reproduction updates."""
+
+from crash_analysis import crash_analyzer
+from crash_analysis.crash_comparer import CrashComparer
+from crash_analysis.stack_parsing import stack_analyzer
+from datastore import data_handler
+from datastore import data_types
+from handlers import base_handler
+from libs import handler
+from libs import helpers
+from metrics import logs
+
+
+def _mark_as_fixed(testcase, revision):
+  """Mark bug as fixed."""
+  testcase.open = False
+  # Bisection not available for external reproduction infrastructure. Assume
+  # range (crash revision : current revision).
+  testcase.fixed = f'{testcase.crash_revision}:{revision}'
+  data_handler.update_progression_completion_metadata(
+      testcase, revision, message=f'fixed in r{revision}')
+
+
+def _mark_errored(testcase, revision, error):
+  """Mark testcase as errored out."""
+  message = 'Received error from external infra, marking testcase as NA.'
+  logs.log_warn(message, error=error, testcase_id=testcase.key.id())
+
+  testcase.fixed = 'NA'
+  testcase.open = False
+  data_handler.update_progression_completion_metadata(
+      testcase, revision, message=message)
+
+
+def handle_update(testcase, revision, stacktrace, error):
+  """Handle update."""
+  logs.log('Got external update for testcase.', testcase_id=testcase.key.id())
+  if error:
+    _mark_errored(testcase, revision, error)
+    return
+
+  last_tested_revision = (
+      testcase.get_metadata('last_tested_revision') or testcase.crash_revision)
+
+  if revision <= last_tested_revision:
+    logs.log_warn(f'Revision {revision} less than previously tested '
+                  f'revision {last_tested_revision}.')
+    return
+
+  fuzz_target = testcase.get_fuzz_target()
+  if fuzz_target:
+    fuzz_target_name = fuzz_target.binary
+  else:
+    fuzz_target_name = None
+
+  state = stack_analyzer.get_crash_data(
+      stacktrace,
+      fuzz_target=fuzz_target_name,
+      symbolize_flag=False,
+      already_symbolized=True)
+  crash_comparer = CrashComparer(state.crash_state, testcase.crash_state)
+  if not crash_comparer.is_similar():
+    logs.log(
+        'State no longer similar.',
+        testcase_id=testcase.key.id(),
+        old_state=testcase.crash_state,
+        new_state=state.crash_state)
+    _mark_as_fixed(testcase, revision)
+    return
+
+  is_security = crash_analyzer.is_security_issue(
+      state.crash_stacktrace, state.crash_type, state.crash_address)
+  if is_security != testcase.security_flag:
+    logs.log('Security flag no longer matches.', testcase_id=testcase.key.id())
+    _mark_as_fixed(testcase, revision)
+    return
+
+  logs.log('Still crashes.', testcase_id=testcase.key.id())
+  testcase.last_tested_crash_stacktrace = stacktrace
```

Comment:
Do we need to refetch testcase to avoid race updates. Same for some branches like line 88, line 81

Reply:
Shouldn't be needed. Unlike native jobs these don't run any other tasks and this is the only place updates happen.